connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfigTransformer.java

```diff
@@ -81,7 +81,7 @@ private void scheduleReload(String connectorName, String path, long ttl) {
         Map<String, HerderRequest> connectorRequests = requests.get(connectorName);
         if (connectorRequests == null) {
             connectorRequests = new ConcurrentHashMap<>();
-            requests.put(connectorName, connectorRequests);
+            requests.putIfAbsent(connectorName, connectorRequests);
```

Comment:
This is still incorrect. Consider the following scenario:

Thread T1 arrives and reaches line 83. Context switch occurs to thread T2.
T2 arrives and also reaches line 83. Context switch occurs to thread T1.
T1 executes line 99 and adds an entry. Context switch occurs to thread T2.
In a world where we want serialized execution, T2 should have gone to the else block, cancelled the previous request and added the new request but now, T2 will simply overwrite the request at line 99 without having cancelled the previous one.

Reply:
You should probably update this code as follows:
```
private void scheduleReload(String connectorName, String path, long ttl) {
        final Map<String, HerderRequest> connectorRequests = requests.computeIfAbsent(connectorName, (c) -> new ConcurrentHashMap<>());
        connectorRequests.computeIfPresent(path, (p, previousRequest) -> {
            // Delete previous request for ttl which is now stale
            previousRequest.cancel();

            log.info("Scheduling a restart of connector {} in {} ms", connectorName, ttl);
            Callback<Void> cb = (error, result) -> {
                if (error != null) {
                    log.error("Unexpected error during connector restart: ", error);
                }
            };
            return worker.herder().restartConnector(ttl, connectorName, cb);
        });
    }
```

Reply:
You might want to use code similar to the following:
```
    private void scheduleReload(String connectorName, String path, long ttl) {
        final Map<String, HerderRequest> connectorRequests = requests.computeIfAbsent(connectorName, (c) -> new ConcurrentHashMap<>());
        connectorRequests.compute(path, (p, previousRequest) -> {
            if (previousRequest != null) {
                // Delete previous request for ttl which is now stale
                previousRequest.cancel();
            }

            log.info("Scheduling a restart of connector {} in {} ms", connectorName, ttl);
            Callback<Void> cb = (error, result) -> {
                if (error != null) {
                    log.error("Unexpected error during connector restart: ", error);
                }
            };
            return worker.herder().restartConnector(ttl, connectorName, cb);
        });
    }
```